### ISSUE TITLE
Group-2: Data Preprocessing with Sklearn Pipeline, Imputation, Feature Encoding & Label Imputation

### ISSUE THREAD

**Issue by project-delphi** [ccm-bioinfo/Camda24_resistance#6]
### 

**Objective**:
Preprocess the dataset by handling missing values, encoding features, imputing missing labels, and ensuring the entire process is structured within an `sklearn` pipeline.

#### Tasks to Complete

1. **Imputation (Handling Missing Data)**
   - [x] Identify features with missing values.
   - [x] Decide on an imputation strategy:
     - [x] Numerical features: Mean/median/mode or advanced methods (e.g., KNNImputer).
     - [x] Categorical features: Mode imputation or adding a "missing" category.
   - [x] Implement imputation using `sklearn`'s `SimpleImputer` or other suitable transformers.
   - [x] Validate that missing values are correctly handled within the pipeline.

2. **Feature Encoding**
   - [x] Identify categorical features.
   - [x] Choose encoding method:
     - [x] One-Hot Encoding for nominal features using `OneHotEncoder`.
     - [x] Ordinal Encoding for ordinal features using `OrdinalEncoder`.
   - [x] Add feature encoding into the pipeline and verify the output is as expected.

3. **Label Imputation**
   - [x] Check for missing labels in the target column.
   - [x] Implement label imputation strategy (e.g., most frequent class or predictive imputation).
   - [x] Add label imputation to the pipeline and ensure correct implementation.

4. **Label Encoding**
   - [x] Ensure labels are in the correct format for modeling.
   - [x] If necessary, apply `LabelEncoder` to transform labels into numerical format.
   - [x] Add label encoding into the pipeline.

5. **Create and Organize Sklearn Pipeline**
   - [ ] Organize all preprocessing steps into an `sklearn.pipeline.Pipeline`.
   - [ ] Ensure pipeline steps are in the correct order:
     1. Feature imputation.
     2. Feature encoding.
     3. Label imputation.
     4. Label encoding.
   - [ ] Test the pipeline on sample data and ensure transformations are applied correctly.

#### Code Structure Example

Ensure the pipeline code is well-organized. Example structure:

```python
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, LabelEncoder
from sklearn.compose import ColumnTransformer

# Example numerical and categorical features
numerical_features = ['feature1', 'feature2']
categorical_features = ['feature3', 'feature4']

# Pipeline for numerical and categorical preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', SimpleImputer(strategy='mean'), numerical_features),
        ('cat', OneHotEncoder(), categorical_features)
    ])

# Main pipeline
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    # Add label imputation and encoding here
])

# Fit pipeline
pipeline.fit(X_train, y_train)

**Comment by project-delphi** [comment]
reset to clean modules with tested code.
